### PR TITLE
Added group `diag`

### PR DESCRIPTION
**Pull request by RichardLitt**

### CODE REVIEW

File: apiary.apib [hunk 1]
```diff
@@ -972,10 +972,309 @@ Outputs the list of peers that were removed.
...
-## sys
+## net [GET /diag/net{?vis}]
+Sends out a message to each node in the network recursively
+requesting a listing of data about them including number of
+connected peers and latencies between them.
+
+The given timeout will be decremented 2s at every network hop,
+ensuring peers try to return their diagnostics before the initiator's
+timeout. If the timeout is too small, some peers may not be reached.
+30s and 60s are reasonable timeout values, though network vary.
```
Comment (dignifiedquire) [inline apiary.apib]: "though network vary" sounds wrong

File: apiary.apib [hunk 2]
```diff
@@ -972,10 +972,309 @@ Outputs the list of peers that were removed.
...
+The default timeout is 20 seconds.
+
++ Parameters
+    + vis (enum[string], optional)
```
Comment (dignifiedquire) [inline apiary.apib]: Is timeout not a parameter?

**Comment by RichardLitt** [comment]
@whyrusleeping Need your feedback on this. These need to be changed.

**Comment by daviddias** [comment]
Note: 0.4.0 has an extra subcommand to ipfs diag

`ipfs diag cmds - List commands run on this ipfs node`

**Comment by RichardLitt** [comment]
@RichardLitt Timeout is global: find the relevant go-code and put it in a thing.

File: apiary.apib (hunk 1 above)
Comment (RichardLitt) [inline apiary.apib]: https://github.com/ipfs/go-ipfs/pull/2486

File: apiary.apib (hunk 2 above)
Comment (RichardLitt) [inline apiary.apib]: No, it is global. I'll need to add a note about that, but it isn't actually relevant to this PR.

**Comment by RichardLitt** [comment]
Added subcmds. Also edited them, so, this PR should be merged first. https://github.com/ipfs/go-ipfs/pull/2487

**Comment by RichardLitt** [comment]
Final thing to do: Add a PR for timeout. However, the rest of this is RFCR and then RFM.

**Comment by RichardLitt** [comment]
Mentioned PR in https://github.com/ipfs/go-ipfs/issues/2583. Merging this as that isn't relevant to the direct code in this PR.

**Pull request merged**